llama-index-core/llama_index/core/instrumentation/__init__.py

```diff
@@ -31,3 +39,38 @@ def get_dispatcher(name: str = "root") -> Dispatcher:
     )
     root_manager.add_dispatcher(new_dispatcher)
     return new_dispatcher
+
+
+class DispatcherSpanMixin(ABC):
+    """
+    Apply dispatcher.span decorator to implementations of abstract methods
+    as well as any previously decorated methods that have been overridden.
+    """
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
+        super().__init_subclass__(**kwargs)
+        abstract_methods: List[str] = []
+        decorated_methods: List[str] = []
+        for base_cls in inspect.getmro(cls):
+            if base_cls is cls:
+                continue
+            for attr, method in base_cls.__dict__.items():
+                if not callable(method):
+                    continue
+                if (
+                    hasattr(method, "__isabstractmethod__")
+                    and method.__isabstractmethod__
+                ):
+                    abstract_methods.append(attr)
+                elif hasattr(method, DISPATCHER_SPAN_DECORATED_ATTR):
+                    decorated_methods.append(attr)
+        dispatcher = get_dispatcher(cls.__module__)
+        for attr, method in cls.__dict__.items():
+            if (
+                not callable(method)
+                or hasattr(method, "__isabstractmethod__")
+                and method.__isabstractmethod__
+            ):
+                continue
+            if attr in abstract_methods or attr in decorated_methods:
+                setattr(cls, attr, dispatcher.span(method))
```

Comment:
are we decorating items more than once? Or why do we check for the `DISPATCHER_SPAN_DECORATED_ATTR` ? I initially thought it would be for filtering

Reply:
This is because in a user-defined class, the user can still manually decorate the method, in which case this automated decoration would be redundant, so adding and checking `DISPATCHER_SPAN_DECORATED_ATTR` prevents duplicated spans.

Reply:
The check on line 65 is for finding a decorated method in the base class that has been overridden (by `cls`), so that's a separate use-case.

Reply:
I think I'm following, though will take a look at this again tomorrow! Perhaps tho we should add a brief comment here and throughout this function.

Reply:
I added unit tests that hopefully can clarify the intent.